tests/federation/test_federation_sender.py

```diff
@@ -173,17 +173,23 @@ def default_config(self):
         return c
 
     def prepare(self, reactor, clock, hs):
-        # stub out `get_rooms_for_user` and `get_users_in_room` so that the
+        test_room_id = "!room:host1"
+
+        # stub out `get_rooms_for_user` and `get_current_hosts_in_room` so that the
         # server thinks the user shares a room with `@user2:host2`
         def get_rooms_for_user(user_id):
-            return defer.succeed({"!room:host1"})
+            return defer.succeed({test_room_id})
 
         hs.get_datastores().main.get_rooms_for_user = get_rooms_for_user
 
-        def get_users_in_room(room_id):
-            return defer.succeed({"@user2:host2"})
+        async def get_current_hosts_in_room(room_id):
+            if room_id == test_room_id:
+                return ["host2"]
+
+            assert "We only expect our stubbed `get_current_hosts_in_room` to be called with "
+            f"test_room_id={test_room_id} but it was called with room_id={room_id}"
```

Comment:
This doesn't do what I intended and just always passes. Looking for something like [`assert.fail([message])`](https://nodejs.org/api/assert.html#assertfailmessage)

I could manually raise an `AssertionError` 🤔

Reply:
Looks like you want to use `self.fail("message")` here? Where "fail" is https://docs.python.org/3/library/unittest.html?highlight=unittest#unittest.TestCase.fail

Reply:
Otherwise:
 -  you could `assert False, "message"` (at present you are asserting the truthiness of a string and nonempty strings are truthy)
 - We sometimes manually raise a RuntimeError within Synapse itself for this sort of thing

Reply:
Thanks @DMRobertson! Is there a way to get the test to fail with the message provided? `self.fail(...)` ends up as the same generic message.

Maybe need to fail it immediately somehow instead of the errors building up and it giving the `twisted.trial.unittest.FailTest: 0 != 3` message instead. I'd like the test to be clear why it failed.

Reply:
Ohh, woops, that message is from,

```
  File "/Users/eric/Documents/github/element/synapse/tests/federation/test_federation_sender.py", line 382, in test_delete_devices
    self.assertEqual(len(self.edus), 3)
```

But I think the `self.fail(...)` is what's making it fail in the first place but not applying.

Reply:
Huh, it looks like `TestCase.fail` just [raises an assertion error by default](https://github.com/python/cpython/blob/3.10/Lib/unittest/case.py#L673). Who knew?

In both cases I'm guessing that the assertion error is caught within Synapse and that's why you're not seeing the test fail.

Assuming that something in Synapse is just catching plain `Exception`s, you could try raising a `BaseException` instance. But that's a) really naughty (see [here](https://docs.python.org/3/library/exceptions.html?highlight=baseexception#BaseException) for why) and b) might not play well with the test framework.

Reply:
>  I just get a generic error message:

Note that you're getting an error message from a _different_ assertion failure! In this example, `self.assertEqual(len(self.edus), 3)`---which I'm assuming is asserted after the stubbed function is called internally

Reply:
Sucks that `TestCase.fail` doesn't do something a bit more encompassing.

Can we make a `TestException` and then adjust some handlers to not be as greedy? Too much maintenance and custom?

```
try:
    something
except TestException:
    raise
except Exception:
    # handling
```

Reply:
Maybe some callback that can fail the test somehow?

Maybe in an after hook, we can check for an error that was set by a custom `self.fail` we override in `HomeserverTestCase`

Reply:
I've removed the assertion for now and left a TODO to tackle this in the future ⏩